### ISSUE TITLE
About NTK-RoPE

### ISSUE THREAD

**Issue by drx-code** [ziqipang/RandAR#2]
Thank you for your work! In the paper I notice you use NTK-RoPE to generate high resolution images. However , I cannot find NTK_Rope in th released code. So how to use the NTK-RoPE? Thank you!

**Comment by ziqipang** [comment]
@drx-code I plan to release this before January, along with some trained checkpoints, so people can directly check the results. If you are in a hurry of trying this out, a uncleaned version of my code looks like this, for computing the RoPE.

```python
def precompute_freqs_cis_2d_extrapolation(
    grid_size: int, n_elem: int, base: int = 10000, cls_token_num=1, extrapolation_factor: float = 2.0
):
    # split the dimension into half, one for x and one for y
    half_dim = n_elem // 2
    freqs = 1.0 / (
        base ** (torch.arange(0, half_dim, 2)[: (half_dim // 2)].float() / half_dim)
    )
    base_max_pos = int(grid_size)
    # base_max_pos = int(grid_size / extrapolation_factor)
    t = torch.linspace(0, base_max_pos - 1, grid_size, device=freqs.device)
    freqs_extra = torch.outer(t, freqs)  # (real_grid_size, head_dim // 2)

    base_max_pos = int(grid_size // extrapolation_factor)
    t = torch.linspace(0, base_max_pos - 1, grid_size, device=freqs.device)
    freqs_inter = torch.outer(t, freqs)  # (real_grid_size, head_dim // 2)

    freqs = freqs_extra
   # NOTE: This is the key step
    freqs_extra[:, 2:] = freqs_inter[:, 2:]

    freqs_grid = torch.concat(
        [
            freqs[:, None, :].expand(-1, grid_size, -1),
            freqs[None, :, :].expand(grid_size, -1, -1),
        ],
        dim=-1,
    )  # (grid_size, grid_size, head_dim // 2)
    cache_grid = torch.stack(
        [torch.cos(freqs_grid), torch.sin(freqs_grid)], dim=-1
    )  # (grid_size, grid_size, head_dim // 2, 2)
    cache = cache_grid.flatten(0, 1)
    cond_freqs = torch.zeros(cls_token_num, n_elem // 2, 2)
    cond_freqs[0, :] = cache[0, :]
    cond_cache = torch.cat((cond_freqs, cache), dim=0)
    return cond_cache
```

A minor clarification: our original quote from the paper is

> In this step, we use top-k high frequency components in the extrapolated RoPE to replace the interpolated RoPE, which better captures the finer details, motivated by NTK-RoPE [12] for extending context lengths in language models.

So we are not using the NTK-RoPE (I tried it in our case, not working really well) but motivated by its design.

Please let me know if you have further questions.

**Comment by ziqipang** [comment]
@drx-code Yes, your understanding is largely correct. Let me clarify again for verification. (1) In the training stage, we use the regular `precompute_freqs_cis_2d` in [here](https://github.com/ziqipang/RandAR/blob/c0ea22f2898388816c13415717c2dc71cd1386e0/RandAR/model/llamagen_gpt.py#L670). (2) During the resolution extrapolation, we will use the two-stage RoPE

>  How to set RoPE in the two stages?

We do this by (1) controlling the order of the tokens; (2) controlling the RoPE. 

* Token Order: The default fully random order is implemented [here](https://github.com/ziqipang/RandAR/blob/c0ea22f2898388816c13415717c2dc71cd1386e0/RandAR/model/randar_gpt.py#L524). We will put the odd-indices like (2x, 2y) first for resolution extrapolation.
* RoPE: when generating the (2x, 2y) tokens, we will use the `freqs_inter` as in the function above. Then for the remaining tokens, we will use the RoPE of `freqs_extra` in the above function. (I think I made a mistake when copying this function, `freqs=freqs_extra` should be after `freqs_extra[:, 2:] = freqs_inter[:, 2:]`, I just corrected this in the thread above.)

**Comment by drx-code** [comment]
Thank you! I'll try it

2025-01-14 01:36:52 "Ziqi Pang" ***@***.***> 写道：

@drx-code I have updated the resolution extrapolation code. Please try it out.
